sdk/tests/conformance/extensions/webgl-polygon-mode.html

```diff
@@ -0,0 +1,184 @@
+    shouldBeNull("gl.getParameter(0x2A02 /* POLYGON_OFFSET_LINE_WEBGL */)");
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "parameter unknown without enabling the extension");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+
+    debug("Check the cap without the extension");
+    gl.disable(0x2A02 /* POLYGON_OFFSET_LINE_WEBGL */);
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "cap unknown without enabling the extension");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+
+    gl.enable(0x2A02 /* POLYGON_OFFSET_LINE_WEBGL */);
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "cap unknown without enabling the extension");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+
+    shouldBeFalse("gl.isEnabled(0x2A02 /* POLYGON_OFFSET_LINE_WEBGL */)");
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "cap unknown without enabling the extension");
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+}
+
+function checkEnums() {
+    debug("");
+    debug("Check enums");
+    shouldBe("ext.POLYGON_MODE_WEBGL", "0x0B40");
+    shouldBe("ext.POLYGON_OFFSET_LINE_WEBGL", "0x2A02");
+    shouldBe("ext.LINE_WEBGL", "0x1B01");
+    shouldBe("ext.FILL_WEBGL", "0x1B02");
+}
+
+function checkQueries() {
+    debug("");
+    debug("Check default state");
+    shouldBe('gl.getParameter(ext.POLYGON_MODE_WEBGL)', 'ext.FILL_WEBGL');
+    shouldBeFalse('gl.getParameter(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    shouldBeFalse('gl.isEnabled(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    debug("");
+    debug("Check state updates");
+    ext.polygonModeWEBGL(gl.FRONT_AND_BACK, ext.LINE_WEBGL);
+    shouldBe('gl.getParameter(ext.POLYGON_MODE_WEBGL)', 'ext.LINE_WEBGL');
+    ext.polygonModeWEBGL(gl.FRONT_AND_BACK, ext.FILL_WEBGL);
+    shouldBe('gl.getParameter(ext.POLYGON_MODE_WEBGL)', 'ext.FILL_WEBGL');
+    debug("");
+    debug("Check errors");
+    ext.polygonModeWEBGL(gl.FRONT, ext.LINE_WEBGL);
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "invalid face");
+    ext.polygonModeWEBGL(gl.FRONT_AND_BACK, 0);
+    wtu.glErrorShouldBe(gl, gl.INVALID_ENUM, "invalid mode");
+    shouldBe('gl.getParameter(ext.POLYGON_MODE_WEBGL)', 'ext.FILL_WEBGL');
+    debug("");
+    debug("Check cap updates");
+    gl.enable(ext.POLYGON_OFFSET_LINE_WEBGL);
+    shouldBeTrue('gl.getParameter(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    shouldBeTrue('gl.isEnabled(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    gl.disable(ext.POLYGON_OFFSET_LINE_WEBGL);
+    shouldBeFalse('gl.getParameter(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    shouldBeFalse('gl.isEnabled(ext.POLYGON_OFFSET_LINE_WEBGL)');
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR, "should be no errors");
+}
+
+function checkDiagonal(r, g, b) {
+    const pixels = new Uint8Array(w * h * 4);
+    gl.readPixels(0, 0, w, h, gl.RGBA, gl.UNSIGNED_BYTE, pixels);
+    for (let i = 0; i < w; i++)
+    {
+        const p = (i * w + i) * 4;
+        const c = pixels.slice(p, p + 4);
```

Comment:
As a follow-up, this will allocate a lot of temporary Uint8Array objects. To optimize this would you consider calling `p` the `baseIndex`, or similar, and checking `pixels[baseIndex + 0]`, `+ 1`, and `+ 2`?

Reply:
Fixed in #3574.